### DISCUSSION TITLE
How does renovate handle multiple docker registries where each needs its own remote / mirror registry?

### DISCUSSION THREAD

**Discussion by alexander-bruun** [renovatebot/renovate#16847] [Request Help] [answered]
I have a kubernetes repository that contains images from various image registries that utilize mirroring underneath the hood for each registry.

This means I'm not able to directly define the full image path in my .yaml definitions.

So far I have setup `hostRules` in my `config.js` for blocking the public image registry, and allowing the private mirror image registry.

```js
hostRules: [
    {
        matchHost: "quay.io",
        enabled: false
    },
    {
        matchHost: "docker.io",
        enabled: false
    },
    {
        matchHost: "https://my-quay-mirror.registry.com",
        username: process.env.REGISTRY_USERNAME,
        password: process.env.REGISTRY_PASSWORD,
    },
    {
        matchHost: "https://my-docker-mirror.registry.com",
        username: process.env.REGISTRY_USERNAME,
        password: process.env.REGISTRY_PASSWORD,
    }
]
```

And I then tried to configure the `default.json` config to contain a set of registryUrl's for the docker datasource that would be shared for all `renovate.json`.

I have tried 2 approaches, but neither worked...

Option 1)

```json
{
    "packageRules": [
        {
            "matchDatasources": [
                "docker"
            ],
            "registryUrls": [
                "https://my-quay-mirror.registry.com",
                "https://my-docker-mirror.registry.com"
            ]
        }
    ]
}
```

This is a bad idea in practice since renovate will then go through all registries to make lookups, this would not be good if I then add a third private repo which doesn't contains public images.

This could potentially leak lookups of private images.

And this approach doesn't work in practice due to `WARN: Excess registryUrls found for datasource lookup - using first configured only`, here only docker.io lookups will work, and quay will be ignored. (but the docs says the variable is an array)

Option 2)

```json
{
    "packageRules": [
        {
            "matchDatasources": [
                "docker"
            ],
            # Here i tried the following: matchSourceUrls, matchSourceUrlPrefixes, matchPackageNames and matchPackagePatterns... but none would allow me to individually govern which image will go to which mirror registry.
            "registryUrls": [
                "https://my-quay-mirror.registry.com"
            ]
        },
        {
            "matchDatasources": [
                "docker"
            ],
            # Here i tried the following: matchSourceUrls, matchSourceUrlPrefixes, matchPackageNames and matchPackagePatterns... but none would allow me to individually govern which image will go to which mirror registry.
            "registryUrls": [
                "https://my-docker-mirror.registry.com"
            ]
        }
    ]
}
```

This approach would never create any pull requests since it probably couldn't find any datasource due to bad configuration.

Which leads to the actual question... am I overlooking some simple way to configure **MULTIPLE** docker registries where each can have their own unique mirror?

**Comment by secustor** [comment, marked as answer]
:thinking:     
I think there is no straight forward way to overwrite the registryUrls based on the original registryUrl currently, but a `matchRegistryURL` parameter should allow your second option.   
If you are interested in that feature you can open a feature request.

**Reply by alexander-bruun** [reply]
That would be very cool! And I believe it would solve my usecase.

I created a feature request here: https://github.com/renovatebot/renovate/issues/16855 I hope the explanation was coherent and understandable.

Thanks for the quick response :)
